rust/otap-dataflow/crates/otap/src/tls_utils.rs

```diff
@@ -0,0 +1,696 @@
+        let acceptor = tls_acceptor.clone();
+        async move {
+            match conn_res {
+                Ok(conn) => {
+                    // Try TLS handshake
+                    match acceptor.accept(conn).await {
+                        Ok(stream) => Some(Ok::<_, io::Error>(stream)),
+                        Err(e) => {
+                            // TLS handshake failed - log and continue
+                            log::warn!("TLS handshake failed: {}", e);
+                            None
+                        }
+                    }
+                }
+                Err(e) => {
+                    // Transport-level listener error - propagate to terminate server
+                    Some(Err(e))
+                }
+            }
+        }
+    })
+}
+
+/// A certificate resolver that lazily reloads TLS certificates with throttled file modification time (mtime) checks.
+///
+/// # Performance characteristics
+///
+/// - Mtime checks for the certificate and key files are throttled by a configurable interval (`check_interval_secs`)
+///   to minimize filesystem operations. This means the filesystem is not polled on every handshake, but only after
+///   the specified interval has elapsed since the last check.
+/// - If a file change is detected during a TLS handshake (i.e., the mtime has changed since the last check),
+///   the certificate and key are reloaded asynchronously. This ensures that new certificates are picked up without
+///   requiring a server restart, while avoiding excessive filesystem access.
+/// - Reloads happen in the context of TLS handshakes. Certificate updates are applied to subsequent connections after
+///   the async reload completes, which typically happens within a short time after file modification is detected.
+#[derive(Debug)]
+pub struct LazyReloadableCertResolver {
+    /// Current certificate
+    cert_key: Arc<ArcSwap<CertifiedKey>>,
+
+    /// File paths
+    cert_path: PathBuf,
+    key_path: PathBuf,
+
+    /// Last known modification times (Arc for sharing with async reload tasks)
+    cert_mtime: Arc<AtomicU64>,
+    key_mtime: Arc<AtomicU64>,
+
+    /// Last time we checked mtime (unix timestamp in seconds)
+    last_check_time: AtomicU64,
+
+    /// Minimum interval between mtime checks (seconds)
+    check_interval_secs: u64,
+
+    /// Reload lock (Arc for sharing with async reload tasks)
+    is_reloading: Arc<AtomicBool>,
+}
+
+impl LazyReloadableCertResolver {
+    /// Creates a new LazyReloadableCertResolver
+    pub fn new(
+        cert_path: PathBuf,
+        key_path: PathBuf,
+        check_interval: Option<Duration>,
+    ) -> Result<Self, io::Error> {
+        let cert_key = load_certified_key_sync(&cert_path, &key_path)?;
+        let cert_mtime = get_mtime(&cert_path)?;
+        let key_mtime = get_mtime(&key_path)?;
+        let now = current_timestamp();
+
+        Ok(Self {
+            cert_key: Arc::new(ArcSwap::from_pointee(cert_key)),
```

Comment:
TIL: arc_swap meant for this kind of long-lived periodically updated runtime and/or configuration struct. https://docs.rs/arc-swap/latest/arc_swap/